### PR TITLE
[pigeon] Fix missed line break between generated statements in Kotlin

### PR DESCRIPTION
**Pull request by ycherniavskyi**
Kotlin generator creates source code with a compiler error, as described in https://github.com/flutter/flutter/issues/118012.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by gaaclarke — CHANGES_REQUESTED** [review]
Nice catch!   Thanks.  LGTM except the test.  It isn't actually testing the fix.  I think ideally we'd have a test that actually compiles the kotlin code.

@tarrinneal can correct me if I'm wrong but I think your best bet is to add a test to the platform_tests.  Here are some for kotlin already: https://github.com/flutter/packages/tree/main/packages/pigeon/platform_tests/test_plugin/android/src/test/kotlin/com/example/test_plugin

**Comment by ycherniavskyi** [comment]
@gaaclarke I don't quite agree that test is not testing the fix. The issue was in the absence of a new line after the assigned statement, and the next expectations test exactly that (`\n` in the end contains string):
```dart
    expect(code, contains('val field1 = Foo.ofRaw(list[0] as Int)!!\n'));
    expect(code, contains('val field2 = list[1] as String\n'));
```
But probably we could join these two separate expects in one, something like this:
```dart
    expect(
      code,
      matches(
        r'val field1 = Foo\.ofRaw\(list\[0\] as Int\)!!\n\s*val field2 = list\[1\] as String\n',
      ),
    );
```

As for platform_tests, if it also needs to be extended with this issue case, then it will be great to have some help/guidance/example on how to do this correctly.

**Review by tarrinneal — APPROVED** [review]
This seems fine to me with the current tests. I will let @gaaclarke have final say on that though. I'm not sure adding a platform test for a line break is necessary.

**Comment by gaaclarke** [comment]
The fact that this doesn't compile indicates we have a gap in our tests that compile pigeon output (platform_tests).  Considering the input is pretty pedestrian I think we really should address it.  The unit tests checks that the newline is there, but a platform_test will assert that the construct in pigeon actually compiles.

The enum test seems the most appropriate:
https://github.com/flutter/packages/blob/main/packages/pigeon/platform_tests/test_plugin/android/src/test/kotlin/com/example/test_plugin/EnumTest.kt

The input pigeon to this test is here: https://github.com/flutter/packages/blob/main/packages/pigeon/pigeons/enum.dart

The minimum thing we should do is edit that pigeon file since that alone will catch the compilation problem.  I'm not sure if more has to be asserted beyond the addition there.  That's probably sufficient.

@tarrinneal at some point we should consider a class of tests that actually execute compilation for fixes like this.  Editing that pigeon will give us coverage but it's hard to associate that was an actual bug that was fixed.

**Comment by tarrinneal** [comment]
> The fact that this doesn't compile indicates we have a gap in our tests that compile pigeon output 

Ah, I missed the compilation failure. Gotta work on my reading comprehension...

**Comment by ycherniavskyi** [comment]
@gaaclarke, extended one of the platform test pigeon file to catch the fixed issue. It is not easy to classify this issue and put it in a particular thematic pigeon file, which is why I put it in a more general one.

<details><summary>Part of the output before the fix</summary>
<p>

```
$ dart run tool/run_tests.dart -t android_kotlin_unittests
...

> Task :test_plugin:compileDebugKotlin FAILED
e: /Users/u/packages/packages/pigeon/platform_tests/test_plugin/android/src/main/kotlin/com/example/test_plugin/AndroidUnittests.gen.kt: (92, 62): Unexpected tokens (use ';' to separate expressions on the same line)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test_plugin:compileDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
58 actionable tasks: 4 executed, 54 up-to-date
```

</p>
</details>

<details><summary>Part of the output after the fix</summary>
<p>

```
$ dart run tool/run_tests.dart -t android_kotlin_unittests
...

BUILD SUCCESSFUL in 4s
67 actionable tasks: 10 executed, 57 up-to-date
```

</p>
</details>

**Comment by stuartmorgan** [comment]
> It is not easy to classify this issue and put it in a particular thematic pigeon file, which is why I put it in a more general one.

Please see https://github.com/flutter/packages/blob/main/packages/pigeon/pigeons/README.md?plain=1#L3-L5

What exactly is the construction that causes this? Is it just any class with an enum field that's not the last field?

If so, the best place to add a compilation test would be to change the order in `AllTypes` and `AllNullableTypes`, with a comment explaining that the enum must not be last with a reference to the issue (also, please file an issue describing the bug to refer to).

**Comment by ycherniavskyi** [comment]
@stuartmorgan, apply your recommendation about the place for extending the compilation test, but instead of moving the enum field around, I think it will be more clear to double it so it exactly shows the issue cause.

here is the test output before the fix:
```
...

> Task :test_plugin:compileDebugKotlin FAILED

58 actionable tasks: 4 executed, 54 up-to-date
e: /Users/user/project/flutter/packages/packages/pigeon/platform_tests/test_plugin/android/src/main/kotlin/com/example/test_plugin/CoreTests.gen.kt: (61, 52): Unexpected tokens (use ';' to separate expressions on the same line)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test_plugin:compileDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
```

**Comment by stuartmorgan** [comment]
> instead of moving the enum field around, I think it will be more clear to double it so it exactly shows the issue cause.

Why? Your own filed issue demonstrates it using one enum that's not last, so I don't see how having one enum that's not last doesn't clearly show the issue.

**Comment by ycherniavskyi** [comment]
@stuartmorgan by moving the enum to not the last position, we reveal the current issue but potentially open the hole for the same issue for other types (in this case, for `Map`).

I just look at this issue more widely and see that the core line that opens the possibility for this issue is:
https://github.com/flutter/packages/blob/3ef0f63290fd48e0fe621d8bd218e8a0646903bb/packages/pigeon/lib/kotlin_generator.dart#L588
and if we remove it, we will reveal the current issue with the enum and other similar position-related issues in the future without even modifying the current compilation test 🤔.

**Comment by stuartmorgan** [comment]
> @stuartmorgan by moving the enum to not the last position, we reveal the current issue but potentially open the hole for the same issue for other types (in this case, for `Map`).

Enums have special codepaths, so are much more likely than most types to have behavior specific to them. Putting something like `String` last is very unlikely to have the same issue.

Also, we have two structures, `AllTypes` and `AllNullableTypes`. We can use different orders in the two (with a comment at the end of each explaining that they should end with different types) to further minimize the risk of an edge case like this slipping through again.

**Comment by ycherniavskyi** [comment]
@stuartmorgan yes, I understand and agree with your point; that is why in my last commit, I reverted the compilation test modification and just removed the core cause of position-related issues - I mean this line:
https://github.com/flutter/packages/blob/3ef0f63290fd48e0fe621d8bd218e8a0646903bb/packages/pigeon/lib/kotlin_generator.dart#L588

**Comment by ycherniavskyi** [comment]
> Also, we have two structures, `AllTypes` and `AllNullableTypes`. We can use different orders in the two (with a comment at the end of each explaining that they should end with different types) to further minimize the risk of an edge case like this slipping through again.

Unfortunately, I am not sure that this could help. Notnullable and Nullable types go in different code paths because of [such conditions](https://github.com/flutter/packages/blob/3ef0f63290fd48e0fe621d8bd218e8a0646903bb/packages/pigeon/lib/kotlin_generator.dart#L551) and as result we don't have this issue with [Nullable enum](https://github.com/flutter/packages/blob/3ef0f63290fd48e0fe621d8bd218e8a0646903bb/packages/pigeon/lib/kotlin_generator.dart#L564) but have it with [Notnullable one](https://github.com/flutter/packages/blob/3ef0f63290fd48e0fe621d8bd218e8a0646903bb/packages/pigeon/lib/kotlin_generator.dart#L580) (the same statement relevant for #3004).

**Comment by stuartmorgan** [comment]
> in my last commit, I reverted the compilation test modification

Reverting the compilation test isn't desirable here. I've described a simple change that would give compilation coverage of this—putting something that isn't a special-case type that has its own codepaths, like String, last.

File: packages/pigeon/CHANGELOG.md [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
 
-* Creates new Generator classes for each language.
+* [kotlin] Fixes a bug with a missed line break between generated statements in the `fromList` function of the companion object.
 
-## 4.2.16
+## 5.0.0
 
-* [swift] Fixes warnings with `Object` parameters.
-* [dart] Fixes warnings with `Object` return values.
-* [c++] Generation of APIs that use `Object` no longer fails.
```
Comment (gaaclarke) [inline packages/pigeon/CHANGELOG.md:9]: You've accidentally deleted the changelog for 4.2.16 =)

**Review by gaaclarke — CHANGES_REQUESTED** [review]

File: packages/pigeon/CHANGELOG.md (hunk 1 above)
Comment (ycherniavskyi) [reply packages/pigeon/CHANGELOG.md]: Ups, restored 😉

**Comment by ycherniavskyi** [comment]
@gaaclarke is there any chance that this PR will be merged?

It is all around this fix:
```diff
--- a/packages/pigeon/lib/kotlin_generator.dart
+++ b/packages/pigeon/lib/kotlin_generator.dart
@@ -592,7 +592,7 @@ void generateKotlin(KotlinOptions options, Root root, StringSink sink) {
                     'val ${field.name} = $fieldType.fromList($listValue as List<Any?>)');
               } else if (!hostDatatype.isBuiltin &&
                   rootEnumNameSet.contains(field.type.baseName)) {
-                indent.write(
+                indent.writeln(
                     'val ${field.name} = $fieldType.ofRaw($listValue as Int)!!');
               } else {
                 indent.writeln('val ${field.name} = $listValue as $fieldType');
```
that even without any tests that were introduced with this PR, differ from all other conditions in this method which all ends with a new line at the end of added text.

**Review by gaaclarke — APPROVED** [review]
lgtm!

**Comment by stuartmorgan** [comment]
This will need the latest `main` merged in and the conflicts resolved, then we can land it.

**Comment by tarrinneal** [comment]
@ycherniavskyi We just landed a rather large set of changes to the structure of the generators. The code you've adjusted is now on line 256. It's unlikely git will handle the merge smoothly, but the change is simple enough I don't expect it will be an issue for you.

**Comment by ycherniavskyi** [comment]
@tarrinneal merged with `main` and resolved appeared conflicts.

**Comment by stuartmorgan** [comment]
This is failing CI:

```
error: argument 'aNullableByteArray' must precede argument 'aNullableString'
      aNullableByteArray: FlutterStandardTypedData(bytes: "1234".data(using: .utf8)!),
~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Looks like some test code needs to be updated for the reordering.
